src/main.rs

```diff
@@ -180,14 +185,143 @@ fn build_memory_map<T: GuestStagePagingMode>(fdt: &Fdt) -> MemMapResult<HwMemMap
     Ok(mem_map)
 }
 
+// Returns the number of PTE pages needed to map all regions in the given memory map.
+// Slightly overestimates of number of pages needed as some regions will share PTE pages in reality.
+fn pte_page_count(mem_map: &HwMemMap) -> u64 {
+    mem_map.regions().fold(0, |acc, r| {
+        acc + Sv48::max_pte_pages(r.size() / PageSize::Size4k as u64)
+    })
+}
+
+// Returns the base address of the first available region in the memory map that is at least `size`
+// bytes long. Returns None if no region is big enough.
+fn find_available_region(mem_map: &HwMemMap, size: u64) -> Option<SupervisorPageAddr> {
+    mem_map
+        .regions()
+        .find(|r| r.region_type() == HwMemRegionType::Available && r.size() >= size)
+        .map(|r| r.base())
+}
+
+// Returns the base, size, and permission pair for the given region if that region type should be
+// mapped in the hypervisor's virtual address space.
+fn hyp_map_params(r: &HwMemRegion) -> Option<(PageAddr<SupervisorPhys>, u64, PteLeafPerms)> {
+    match r.region_type() {
+        HwMemRegionType::Available => {
+            // map available memory as rwx - unser what it'll be used for.
+            Some((r.base(), r.size(), PteLeafPerms::RWX))
+        }
+        HwMemRegionType::Reserved(HwReservedMemType::FirmwareReserved) => {
+            // No need to map regions reserved for firmware use
+            None
+        }
+        HwMemRegionType::Reserved(HwReservedMemType::HypervisorImage)
+        | HwMemRegionType::Reserved(HwReservedMemType::HostKernelImage)
+        | HwMemRegionType::Reserved(HwReservedMemType::HostInitramfsImage) => {
```

Comment:
Shouldn't the host kernel & initramfs be non-executable from within the hypervisor?